components/autofill/src/api/credit_cards.rs

```diff
@@ -0,0 +1,594 @@
+            cc_exp_month                = :cc_exp_month,
+            cc_exp_year                 = :cc_exp_year,
+            cc_type                     = :cc_type,
+            time_created                = :time_created,
+            time_last_used              = :time_last_used,
+            time_last_modified          = :time_last_modified,
+            times_used                  = :times_used,
+            sync_change_counter         = sync_change_counter + 1
+        WHERE guid                      = :guid",
+        rusqlite::named_params! {
+            ":cc_name": credit_card.fields.cc_name,
+            ":cc_number": credit_card.fields.cc_number,
+            ":cc_exp_month": credit_card.fields.cc_exp_month,
+            ":cc_exp_year": credit_card.fields.cc_exp_year,
+            ":cc_type": credit_card.fields.cc_type,
+            ":time_created": credit_card.time_created,
+            ":time_last_used": credit_card.time_last_used,
+            ":time_last_modified": credit_card.time_last_modified,
+            ":times_used": credit_card.times_used,
+            ":guid": credit_card.guid,
+        },
+    )?;
+
+    tx.commit()?;
+    Ok(())
+}
+
+#[allow(dead_code)]
+pub fn delete_credit_card(conn: &mut Connection, guid: &Guid) -> Result<bool> {
+    let tx = conn.transaction()?;
+
+    // check that guid exists
+    let exists = tx.query_row(
+        "SELECT EXISTS (
+            SELECT 1
+            FROM credit_cards_data d
+            WHERE guid = :guid
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM   credit_cards_tombstones t
+                    WHERE  d.guid = t.guid
+                )
+        )",
+        &[guid.as_str()],
+        |row| row.get(0),
+    )?;
+
+    if exists {
+        tx.execute_named(
+            "DELETE FROM credit_cards_data
+            WHERE guid = :guid",
+            rusqlite::named_params! {
+                ":guid": guid.as_str(),
+            },
+        )?;
+
+        tx.execute_named(
+            "INSERT OR IGNORE INTO credit_cards_tombstones (
+                guid,
+                time_deleted
+            ) VALUES (
+                :guid,
+                :time_deleted
+            )",
+            rusqlite::named_params! {
+                ":guid": guid.as_str(),
+                ":time_deleted": Timestamp::now(),
+            },
+        )?;
+    }
+
+    tx.commit()?;
+    Ok(exists)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db::test::new_mem_db;
+
+    #[test]
+    fn test_credit_card_create_and_read() {
+        let mut db = new_mem_db();
+
+        let saved_credit_card = add_credit_card(
+            &mut db,
+            NewCreditCardFields {
+                cc_name: "jane doe".to_string(),
+                cc_number: "2222333344445555".to_string(),
+                cc_exp_month: 3,
+                cc_exp_year: 2022,
+                cc_type: "visa".to_string(), //TODO: turn this into an enum?
```

Comment:
That's a good general question though - what validation should we perform, if any? On one hand, if a new type is introduced, it would seem dumb for us to need a new release just to add a new string to a list of "valid" types. And if we did that, we'd end up with an issue re versions - eg, desktop rejects the new type but mobile allows it, and they sync!

So all in all, I guess I'm voting for this being a string that we don't touch or validate, and just rely on apps to do the right thing. If they discover later they screwed up, they can use our existing API to fix it without our help.

That said though, it's worth adding that link lina shared to a comment where the struct is defined.

Reply:
That makes sense. The need to bump the version with changes to the payment type seemed silly to me as well and it's unlikely that garbage text would end up in the `cc_type` field. I'll keep it a string and remove the TODO. Thanks!